### PR TITLE
[gdaldem] Use GDALArgumentParser

### PR DESCRIPTION
**Pull request by elpaso**
# General

```gdaldem --long-usage
Usage: gdaldem [--help] [--long-usage] [--help-general] {TPI,TRI,aspect,color-relief,hillshade,roughness,slope}

Tools to analyze and visualize DEMs.

Optional arguments:
  -h, --help         Shows short help message and exits. 
  --long-usage       Shows long help message and exits. 
  --help-general     Report detailed help on general options. 

Subcommands:
  TPI               Compute the Topographic Position Index.
  TRI               Compute the Terrain Ruggedness Index.
  aspect            Compute aspect.
  color-relief      Color relief computed from the elevation and a text-based color configuration file.
  hillshade         Compute hillshade.
  roughness         Compute the roughness of the DEM.
  slope             Compute slope.

For more details, consult https://gdal.org/programs/gdaldem.html
```

# Topographic Position Index

```
gdaldem tpi --long-usage
Usage: TPI [--help] [--long-usage] [--help-general]
           [-of <output_format>] [-compute_edges] [-b <value>] [-co <NAME>=<VALUE>]... [--quiet]
           input_dem output_TPI_map

Compute the Topographic Position Index.

Positional arguments:
  input_dem            The input DEM raster to be processed. 
  output_TPI_map       The output raster to be produced. 

Optional arguments:
  -h, --help           Shows short help message and exits. 
  --long-usage         Shows long help message and exits. 
  --help-general       Report detailed help on general options. 
  -of <output_format>  Output format. 
  -compute_edges       Do the computation at raster edges and near nodata values. 
  -b <value>           Select an input band. 
  -co <NAME>=<VALUE>   Creation option(s). [may be repeated]
  -q, --quiet          Quiet mode. No progress message is emitted on the standard output.
```

# Terrain Ruggedness Index

```
gdaldem tri --long-usage
Export manually or use this script as a wrapper for commands!!!!
Usage: TRI [--help] [--long-usage] [--help-general]
           [-alg Wilson|Riley] [-of <output_format>] [-compute_edges] [-b <value>] [-co <NAME>=<VALUE>]... [--quiet]
           input_dem output_TRI_map

Compute the Terrain Ruggedness Index.

Positional arguments:
  input_dem            The input DEM raster to be processed. 
  output_TRI_map       The output raster to be produced. 

Optional arguments:
  -h, --help           Shows short help message and exits. 
  --long-usage         Shows long help message and exits. 
  --help-general       Report detailed help on general options. 
  -alg Wilson|Riley    Choose between Wilson or Riley algorithms. 
  -of <output_format>  Output format. 
  -compute_edges       Do the computation at raster edges and near nodata values. 
  -b <value>           Select an input band. 
  -co <NAME>=<VALUE>   Creation option(s). [may be repeated]
  -q, --quiet          Quiet mode. No progress message is emitted on the standard output. 
```

# Hillshade

```
hillshade --long-usage
Usage: hillshade [--help] [--long-usage] [--help-general]
                 [-alg <Horn|ZevenbergenThorne>] [-z <factor>] [-s <scale>] [-az <azimuth>] [-alt <altitude>]
                 [[-combined]|[-multidirectional]|[-igor]]
                 [-of <output_format>] [-compute_edges] [-b <value>] [-co <NAME>=<VALUE>]... [--quiet]
                 input_dem output_hillshade

Compute hillshade.

Positional arguments:
  input_dem                      The input DEM raster to be processed. 
  output_hillshade               The output raster to be produced. 

Optional arguments:
  -h, --help                     Shows short help message and exits. 
  --long-usage                   Shows long help message and exits. 
  --help-general                 Report detailed help on general options. 
  -alg <Horn|ZevenbergenThorne>  Choose between ZevenbergenThorne or Horn algorithms. 
  -z <factor>                    Vertical exaggeration. 
  -s <scale>                     Ratio of vertical units to horizontal. 
  -az <azimuth>                  Azimuth of the light, in degrees. 
  -alt <altitude>                Altitude of the light, in degrees. 
  -combined                      Use combined shading. 
  -multidirectional              Use multidirectional shading. 
  -igor                          Use multidirectional shading. 
  -of <output_format>            Output format. 
  -compute_edges                 Do the computation at raster edges and near nodata values. 
  -b <value>                     Select an input band. 
  -co <NAME>=<VALUE>             Creation option(s). [may be repeated]
  -q, --quiet                    Quiet mode. No progress message is emitted on the standard output.
```

# Aspect

```
gdaldem aspect  --long-usage
Usage: aspect [--help] [--long-usage] [--help-general]
              [-alg Horn|ZevenbergenThorne] [-trigonometric] [-zero_for_flat] [-of <output_format>] [-compute_edges]
              [-b <value>] [-co <NAME>=<VALUE>]... [--quiet]
              input_dem output_aspect_map

Compute aspect.

Positional arguments:
  input_dem                    The input DEM raster to be processed. 
  output_aspect_map            The output raster to be produced. 

Optional arguments:
  -h, --help                   Shows short help message and exits. 
  --long-usage                 Shows long help message and exits. 
  --help-general               Report detailed help on general options. 
  -alg Horn|ZevenbergenThorne  Choose between ZevenbergenThorne or Horn algorithms. 
  -trigonometric               Express aspect in trigonometric format. 
  -zero_for_flat               Return 0 for flat areas with slope=0, instead of -9999. 
  -of <output_format>          Output format. 
  -compute_edges               Do the computation at raster edges and near nodata values. 
  -b <value>                   Select an input band. 
  -co <NAME>=<VALUE>           Creation option(s). [may be repeated]
  -q, --quiet                  Quiet mode. No progress message is emitted on the standard output.
```

# Color relief 

```
gdaldem color-relief  --long-usage
Usage: color-relief [--help] [--long-usage] [--help-general]
                    [-alpha] [-exact_color_entry] [-nearest_color_entry] [-of <output_format>] [-compute_edges]
                    [-b <value>] [-co <NAME>=<VALUE>]... [--quiet]
                    input_dem color_text_file output_color_relief_map

Color relief computed from the elevation and a text-based color configuration file.

Positional arguments:
  input_dem                The input DEM raster to be processed. 
  color_text_file          Text-based color configuration file. 
  output_color_relief_map  The output raster to be produced. 

Optional arguments:
  -h, --help               Shows short help message and exits. 
  --long-usage             Shows long help message and exits. 
  --help-general           Report detailed help on general options. 
  -alpha                   Add an alpha channel to the output raster. 
  -exact_color_entry       Use strict matching when searching in the configuration file. 
  -nearest_color_entry     Use the RGBA corresponding to the closest entry in the configuration file. 
  -of <output_format>      Output format. 
  -compute_edges           Do the computation at raster edges and near nodata values. 
  -b <value>               Select an input band. 
  -co <NAME>=<VALUE>       Creation option(s). [may be repeated]
  -q, --quiet              Quiet mode. No progress message is emitted on the standard output. 
```

# Roughness

```
gdaldem roughness --long-usage
Usage: roughness [--help] [--long-usage] [--help-general]
                 [-of <output_format>] [-compute_edges] [-b <value>] [-co <NAME>=<VALUE>]... [--quiet]
                 input_dem output_roughness_map

Compute the roughness of the DEM.

Positional arguments:
  input_dem             The input DEM raster to be processed. 
  output_roughness_map  The output raster to be produced. 

Optional arguments:
  -h, --help            Shows short help message and exits. 
  --long-usage          Shows long help message and exits. 
  --help-general        Report detailed help on general options. 
  -of <output_format>   Output format. 
  -compute_edges        Do the computation at raster edges and near nodata values. 
  -b <value>            Select an input band. 
  -co <NAME>=<VALUE>    Creation option(s). [may be repeated]
  -q, --quiet           Quiet mode. No progress message is emitted on the standard output.
```

# Slope

```
gdaldem slope --long-usage
Usage: slope [--help] [--long-usage] [--help-general]
             [-alg Horn|ZevenbergenThorne] [-p] [-s <scale>] [-of <output_format>] [-compute_edges] [-b <value>]
             [-co <NAME>=<VALUE>]... [--quiet]
             input_dem output_slope_map

Compute slope.

Positional arguments:
  input_dem                    The input DEM raster to be processed. 
  output_slope_map             The output raster to be produced. 

Optional arguments:
  -h, --help                   Shows short help message and exits. 
  --long-usage                 Shows long help message and exits. 
  --help-general               Report detailed help on general options. 
  -alg Horn|ZevenbergenThorne  Choose between ZevenbergenThorne or Horn algorithms. 
  -p                           Express slope as a percentage. 
  -s <scale>                   Ratio of vertical units to horizontal. 
  -of <output_format>          Output format. 
  -compute_edges               Do the computation at raster edges and near nodata values. 
  -b <value>                   Select an input band. 
  -co <NAME>=<VALUE>           Creation option(s). [may be repeated]
  -q, --quiet                  Quiet mode. No progress message is emitted on the standard output.
```

### CODE REVIEW

File: apps/gdaldem_lib.cpp [hunk 1]
```diff
@@ -3763,246 +4199,134 @@ GDALDatasetH GDALDEMProcessing(const char *pszDest, GDALDatasetH hSrcDataset,
...
+                    CPLDebug("GDAL", "Error for mode %s: %s",
+                             processingMode.c_str(), e.what());
+                    continue;
+                }
+                catch (std::logic_error &e)
+                {
+                    CPLDebug("GDAL", "Error for mode %s: %s",
+                             processingMode.c_str(), e.what());
+                    continue;
+                }
```
Comment (rouault) [inline apps/gdaldem_lib.cpp:4309]: I would not emit those debug messages, as they are "expected" errors

File: apps/gdaldem_lib.cpp [hunk 2]
```diff
@@ -3170,6 +3172,436 @@ static Algorithm GetAlgorithm(const char *pszProcessing)
...
+            .store_into(psOptionsForBinary->osSrcFilename)
+            .help(_("The input DEM raster to be processed."));
+
+        subCommandSlope->add_argument("output_slope_map")
+            .store_into(psOptionsForBinary->osDstFilename)
+            .help(_("The output raster to be produced."));
+    }
+
+    subCommandSlope->add_argument("-alg")
+        .metavar("<Horn|ZevenbergenThorne>")
```
Comment (rouault) [inline apps/gdaldem_lib.cpp:3347]:
```suggestion
        .metavar("Horn|ZevenbergenThorne")
```

File: apps/gdaldem_lib.cpp [hunk 3]
```diff
@@ -3170,6 +3172,436 @@ static Algorithm GetAlgorithm(const char *pszProcessing)
...
+            .store_into(psOptionsForBinary->osSrcFilename)
+            .help(_("The input DEM raster to be processed."));
+
+        subCommandAspect->add_argument("output_aspect_map")
+            .store_into(psOptionsForBinary->osDstFilename)
+            .help(_("The output raster to be produced."));
+    }
+
+    subCommandAspect->add_argument("-alg")
+        .metavar("<Horn|ZevenbergenThorne>")
```
Comment (rouault) [inline apps/gdaldem_lib.cpp:3400]:
```suggestion
        .metavar("Horn|ZevenbergenThorne")
```

File: apps/gdaldem_lib.cpp [hunk 4]
```diff
@@ -3170,6 +3172,436 @@ static Algorithm GetAlgorithm(const char *pszProcessing)
...
+            .store_into(psOptionsForBinary->osSrcFilename)
+            .help(_("The input DEM raster to be processed."));
+
+        subCommandTRI->add_argument("output_TRI_map")
+            .store_into(psOptionsForBinary->osDstFilename)
+            .help(_("The output raster to be produced."));
+    }
+
+    subCommandTRI->add_argument("-alg")
+        .metavar("<Wilson|Riley>")
```
Comment (rouault) [inline apps/gdaldem_lib.cpp:3500]:
```suggestion
        .metavar("Wilson|Riley")
```

**Comment by rouault** [comment]
can you paste the output of "gdaldem --help" and "gdaldem --long-usage" ?

File: apps/gdal_utils_priv.h [hunk 5]
```diff
@@ -58,11 +58,11 @@ struct GDALInfoOptionsForBinary
...
-    char *pszProcessing;
-    char *pszSrcFilename;
-    char *pszColorFilename;
-    char *pszDstFilename;
-    int bQuiet;
+    std::string osProcessing;
+    std::string osSrcFilename;
+    std::string osColorFilename;
+    std::string osDstFilename;
+    bool bQuiet;
```
Comment (rouault) [inline apps/gdal_utils_priv.h:65]:
```suggestion
    bool bQuiet = false;
```

File: apps/gdaldem_lib.cpp (hunk 1 above)
Comment (elpaso) [reply apps/gdaldem_lib.cpp]: done

File: apps/gdaldem_lib.cpp [hunk 6]
```diff
@@ -3170,6 +3172,436 @@ static Algorithm GetAlgorithm(const char *pszProcessing)
...
+            .store_into(psOptions->bMultiDirectional)
+            .help(_("Use multidirectional shading."));
+
+    subCommandHillshade->add_hidden_alias_for(multidirectionalHillshadeArg,
+                                              "--multidirectional");
+
+    auto &igorHillshadeArg = shadingGroup.add_argument("-igor")
+                                 .flag()
+                                 .store_into(psOptions->bIgor)
+                                 .help(_("Use multidirectional shading."));
```
Comment (rouault) [inline apps/gdaldem_lib.cpp:3322]:
```suggestion
                                 .help(_("Shading which tries to minimize effects on other map features beneath."));
```

**Comment by rouault** [comment]
When done, would you mind reworking commit history, to have commit "ArgParser: is_used_globally()" first, and squash together all other gdaldem specific commits?

**Comment by elpaso** [comment]
> When done, would you mind reworking commit history, to have commit "ArgParser: is_used_globally()" first, and squash together all other gdaldem specific commits?

I've grouped all argparser commits in a separate one.

**Comment by rouault** [comment]
MSVC not happy:
```
 C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.39.33519\include\algorithm(3433,24): error C2220: the following warning is treated as an error [D:\a\gdal\gdal\build\apps\utils_common.vcxproj]
  (compiling source file '../../apps/gdalargumentparser.cpp')
  
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.39.33519\include\algorithm(3433,24): warning C4244: '=': conversion from 'int' to 'char', possible loss of data [D:\a\gdal\gdal\build\apps\utils_common.vcxproj]
  (compiling source file '../../apps/gdalargumentparser.cpp')
  C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.39.33519\include\algorithm(3433,24):
  the template instantiation context (the oldest one first) is
  	D:\a\gdal\gdal\apps\gdalargumentparser.cpp(420,14):
  	see reference to function template instantiation '_OutIt std::transform<std::_String_iterator<std::_String_val<std::_Simple_types<_Elem>>>,std::_String_iterator<std::_String_val<std::_Simple_types<_Elem>>>,int(__cdecl *)(int)>(const _InIt,const _InIt,_OutIt,_Fn)' being compiled
          with
          [
              _OutIt=std::_String_iterator<std::_String_val<std::_Simple_types<char>>>,
              _Elem=char,
              _InIt=std::_String_iterator<std::_String_val<std::_Simple_types<char>>>,
              _Fn=int (__cdecl *)(int)
          ]
  ```

**Pull request merged**